Stat timers

toshok opened the pull request:
Each thread has a stack of timers (which should have automatic storage.)  creating a timer starts it and pushes it onto the stack, and also pauses the currently running one (if there is one) to avoid double counting.  When a timer goes out of scope it's stopped and popped off the stack, and the previous one on the stack is resumed.

kmod commented on src/core/threading.cpp:
```diff
@@ -107,6 +119,15 @@ class ThreadStateInternal {
+        StatTimer* st = StatTimer::getStack();
+        if (st && !st->isPaused())
+            st->pause();
+        stack.next_generator->statTimers = st;
```
is this right?

kmod commented on src/core/stats.h:
```diff
@@ -55,6 +59,59 @@ struct StatCounter {
+    static StatTimer* getStack() { return stack; }
+    static void setStack(StatTimer* s, bool andResume = false) {
+        stack = s;
+        if (andResume && s && s->isPaused()) {
```
why wouldn't the other stack be paused?

there are a couple places we check isPaused, but it seems like we should just assert(paused)

kmod commented on src/codegen/ast_interpreter.cpp:
```diff
@@ -285,6 +285,11 @@ class RegisterHelper {
+#if SLOWPATH_TIMERS
+    static int _stid = Stats::getStatId("zzz_slowpath_astinterpreter_execute_us");
+    StatTimer _st(_stid);
+#endif
```
I would make this into a macro :)

toshok replied:
yeah, that was already on my todo list :)

On Wed, Apr 29, 2015 at 2:20 PM, Kevin Modzelewski <notifications@github.com
> wrote:

> In src/codegen/ast_interpreter.cpp
> <https://github.com/dropbox/pyston/pull/475#discussion_r29382242>:
>
> > @@ -285,6 +285,11 @@ class RegisterHelper {
> >  Value ASTInterpreter::execute(ASTInterpreter& interpreter, CFGBlock* start_block, AST_stmt* start_at) {
> >      threading::allowGLReadPreemption();
> >
> > +#if SLOWPATH_TIMERS
> > +    static int _stid = Stats::getStatId("zzz_slowpath_astinterpreter_execute_us");
> > +    StatTimer _st(_stid);
> > +#endif
>
> I would make this into a macro :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dropbox/pyston/pull/475/files#r29382242>.
>

toshok commented:
Okay, I think this is finally ready to be looked at seriously and merged.

I removed in the ad-hoc StatTimer in `__cxa_throw` and `__cxa_begin_catch`, because even though I can explain it and it should work, it causes ticks to not add up.  It's definitely more accurate, though, so either I can take a look again, or once the unwinder goes in we can add StatTimers there.

I split things up some of the timers bit more but there's still splitting work to be done wrt the `chosen_cf` timers.  I also cleaned up `jit.cpp` such that we log the main time in ticks as a stat counter itself (previously it was always printing it to the console, something that breaks the test runner.)  The CPU MHz estimate is also logged with stats.

Ticks and StatTimers both store their durations in terms of cpu ticks, and we convert (for any stat that starts with `us_`) at Stats::dump() time.  Anything that starts with `us_timer` is also accumulated and printed at the end.  So checking to verify timers are working properly can be done by checking `ticks_in_main` vs. `ticks_all_timers` - the two should always be equal.  Here's an example snippet from the stats dump running `minibenchmarks/django-template.py`:

```
$ ./pyston_release -x -s minibenchmarks/django-template.py
took 64.5ms for last iteration
Stats:
estimated_cpu_mhz: 2799.99810
...
ticks_in_main: 93958186739
us_timer_PyEq: 89934
us_timer_PyHasher: 124069
us_timer_PyLt: 996
us_timer_astInterpretFunction: 71019
us_timer_astinterpreter_execute: 367038
us_timer_astinterpreter_jump_osrexit: 3577
us_timer_boxedmethoddescriptor__call__: 380664
us_timer_caching_parse_file: 107982
us_timer_chosen_cf_body_builtins: 5053399
us_timer_chosen_cf_body_jitted: 2528026
us_timer_compileFunction: 904530
us_timer_cpyton_parsing: 196588
us_timer_gc_collection: 1796842
us_timer_generator_toplevel: 13068
us_timer_getTopPythonFrame: 107349
us_timer_gettraceback: 13866249
us_timer_importFrom: 660
us_timer_importStar: 3820
us_timer_interpreted_module_toplevel: 295
us_timer_main_toplevel: 43979
us_timer_slowpath_augbinop: 4759
us_timer_slowpath_binop: 75327
us_timer_slowpath_callattr: 1094992
us_timer_slowpath_compare: 192531
us_timer_slowpath_createBoxedIterWrapperIfNeeded: 10995
us_timer_slowpath_delattr: 59
us_timer_slowpath_delitem: 295
us_timer_slowpath_getPystonIter: 2531
us_timer_slowpath_getattr: 3884737
us_timer_slowpath_getclsattr: 35
us_timer_slowpath_getglobal: 28497
us_timer_slowpath_getitem: 335760
us_timer_slowpath_len: 22
us_timer_slowpath_nonzero: 163599
us_timer_slowpath_runtimecall: 1683543
us_timer_slowpath_setitem: 8890
us_timer_slowpath_setsattr: 94094
us_timer_slowpath_unaryop: 13
us_timer_slowpath_unboxedLen: 23
us_timer_tp_getattr: 40
us_timer_tp_getattro: 38492
us_timer_tp_setattr: 0
us_timer_tp_setattro: 4259
us_timer_typeNew: 272916
us_timer_visit_print: 3
zzz_gc_alloc_bytes: 0
ticks_all_timers: 93958186739
```

maybe it would be better to make two passes over the stat counters, accumulating total ticks, then using a StatCounter inside Stats::dump() to cause `ticks_all_timers` and `ticks_in_main` to be right next to each other.  might also be worth it to print some sort of error when they don't match.

kmod commented on src/core/stats.cpp:
```diff
@@ -22,9 +22,108 @@
+        pause(at_time);
+    } else {
+        // fprintf (stderr, "WARNING: timer was paused.\n");
+        at_time = _last_pause_time;
```
when would this happen?

toshok replied:
happens at the end of both the main timer (jit.cpp:373) and generatorEntry (generator.cpp:115)

basically there are a couple cases where we want to mark a timer as paused at a specific time, instead of when the destructor is called at approximately the next line.

Not sure the gymnastics are necessary in jit.cpp anymore, since those were there initially to estimate CPU MHz.  The generator one is to ensure that the timers for different stacks are paused/resumed at the same tick.

toshok replied:
ah, turns out we do need the gymnastics in jit.cpp.  without them the ticks don't match up and we get a warning.  sweet.

kmod commented on src/core/stats.cpp:
```diff
@@ -69,10 +193,22 @@ void Stats::dump(bool includeZeros) {
+
+            if (startswith(pairs[i].first, "us_timer"))
+                accumulated_stat_timer_ticks += (*counts)[pairs[i].second];
+        }
```
Could we do something ilke `if (pairs[1].first == "ticks_in_main") ticks_in_main = pairs[i].second;` and then later just check accumulated_stat_timer_ticks against it?

toshok replied:
yup, definitely.

kmod commented on src/runtime/objmodel.cpp:
```diff
@@ -125,6 +125,7 @@ static Box* (*callattrInternal3)(Box*, const std::string*, LookupScope, CallRewr
        * (*)(Box*, const std::string*, LookupScope, CallRewriteArgs*, ArgPassSpec, Box*, Box*, Box*))callattrInternal;
 
 size_t PyHasher::operator()(Box* b) const {
+    STAT_TIMER(t0, "us_timer_PyHasher");
```
should these be SLOWPATH_TIMER's instead?

kmod commented on src/runtime/generator.cpp:
```diff
@@ -86,28 +86,34 @@ Context* getReturnContextForGeneratorFrame(void* frame_addr) {
+        g->entryExited = true;
+        threading::popGenerator();
+
+        g->timer_time = getCPUTicks(); // store off the timer that our caller (in g->returnContext) will resume at
```
We'll probably have to guard this (and some other stuff in this file) with `#if SLOWPATH_TIMERS`.

toshok replied:
I changed the generator ifdefs from `#if SLOWPATH_TIMERS` to `#if !DISABLE_STATS` since in this particular case we want it to always be there - the timer registered in generatorEntry is the toplevel timer for the generator (much like the main timer in jit.cpp)

kmod commented on src/core/stats.h:
```diff
@@ -22,23 +22,57 @@
+#else
+#define SLOWPATH_TIMER(id, name)
+#define SLOWPATH_TIMER2(id, name, at_time)
+#endif
```
just a thought, but should we combine SLOWPATH_TIMER and STAT_TIMER?  some of the timers we probably want all the time (as opposed to the slowpath ones which seem to add some overhead), but it seems like it's tricky to remember to call the right type of timer.

toshok replied:
agreed.  I think combining them would be good.  I'm running into a problem where I have nearly too many timers to show on a single page of console output.  At some point we might want to have timer levels of some sort, and we can at that point add another parameter to the `STAT_TIMER`.  we could either disable them altogether at compile time based on the level, or keep them in and prune the output based on it.

kmod replied:
hmm thinking about it some more, I think we still want all of our previous timers to run even if we turn the expensive timers off.  maybe we could un-convert some of the ones you converted over?

kmod replied:
oh, and we definitely don't want to have to turn off all stats to get the new stat timers :)  we should probably have the stat timers turned off by default or measure the impact on the benchmark suite.

kmod commented:
lgtm with some minor comments.  we should make sure it works when we do `#define SLOWPATH_TIMERS 0`.

we have some bikeshedding to do on the names of the stats :P  but we can always iterate once it's in :)

toshok commented:
oops.  let's merge this (after travis-ci rebuilds) instead of closing it.
